### ISSUE TITLE
.bss section is accumulated in ELF file

### ISSUE THREAD

**Issue by deepa2015** [llvm/llvm-project#78261]
Hello,
With lld 18.0.0, it is seen that the ELF section file offset computation is incorrect for sections processed after uninitialized section (.bss).
The uninitialized section (.bss) also gets accumulated in raw ELF file which ideally should not.

### Details:
Consider test
 bss-test-arr.c

```c
int arr[1000];
int main()
{
return 0;
}
```

### Linker description file(rep.lds)

```
SECTIONS {
  .rodata : {}
  .text : {}
  .data : {}
  .bss : {}
  .ldata : {
  . = ALIGN(8);
  }
}
```

### Commands executed

```
clang --target=x86_64 -c bss-test-arr.c -o bss-test-arr.o -v
ld.lld -v -m elf_x86_64 -T ./rep.lds bss-test-arr.o -static -o bss-test-arr.out
llvm-readelf -S -l bss-test-arr.out > bss-test-arr.out.txt
```

### Readelf output snippet

Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
..
  [ 3] .bss              NOBITS          0000000000000050 001050 000fa0 00  WA  0   0 16
  [ 4] .ldata            PROGBITS        0000000000000ff0 **001ff0** 000000 00  WA  0   0  1
..

Program Headers:
  Type           Offset   VirtAddr           PhysAddr           FileSiz  MemSiz   Flg Align
..
  LOAD           0x001050 0x0000000000000050 0x0000000000000050 **0x000fa0** 0x000fa0 RW  0x1000
..

The .ldata section file offset should be same as .bss section which is of type NOBITS i.e 0x1050. The program header for .bss also has file size of 0xfa0.

The readelf output is attached for reference.
[bss-test-arr.out.txt](https://github.com/llvm/llvm-project/files/13949235/bss-test-arr.out.txt)

**Comment by MaskRay** [comment]
Yes, we currently have a missing optimization for .bss-style sections as we may waste up to MAXPAGESIZE bytes.
The last paragraph of https://reviews.llvm.org/D150510 mentions this:

> We have a missing optimization to prevent all waste, but implementing it would introduce complexity and likely be error-prone.

I'll dabble around.
